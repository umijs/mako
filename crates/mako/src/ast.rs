pub(crate) mod comments;
pub(crate) mod css_ast;
pub(crate) mod error;
pub mod file;
pub(crate) mod js_ast;
pub(crate) mod sourcemap;
#[cfg(test)]
pub mod tests;
pub(crate) mod utils;
