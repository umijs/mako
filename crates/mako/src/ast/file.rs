use std::hash::Hasher;
use std::io::{BufRead, BufReader};
use std::path::PathBuf;
use std::sync::{Arc, OnceLock};

use mako_core::anyhow::{anyhow, Result};
use mako_core::base64::alphabet::STANDARD;
use mako_core::base64::{engine, Engine};
use mako_core::pathdiff::diff_paths;
use mako_core::regex::Regex;
use mako_core::thiserror::Error;
use mako_core::twox_hash::XxHash64;
use mako_core::{md5, mime_guess};
use percent_encoding::percent_decode_str;
use url::Url;

use crate::compiler::Context;
use crate::utils::base64_decode;

#[derive(Debug, Clone)]
pub struct Asset {
    pub path: String,
    pub content: String,
}

#[derive(Debug, Clone)]
pub struct JsContent {
    pub is_jsx: bool,
    pub content: String,
}

impl Default for JsContent {
    fn default() -> Self {
        JsContent {
            is_jsx: false,
            content: "".to_string(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum Content {
    Js(JsContent),
    Css(String),
    // TODO: unify the assets handler
    // it's used in minifish plugin(bundless mode) only
    // and bundle mode will emit assets to context.assets_info
    Assets(Asset),
}

#[derive(Debug, Error)]
enum FileError {
    #[error("To base64 error: {path:?}")]
    ToBase64Error { path: String },
}

#[derive(Debug, Clone)]
pub struct File {
    pub path: PathBuf,
    pub relative_path: PathBuf,
    pub extname: String,
    pub content: Option<Content>,
    pub is_under_node_modules: bool,
    pub is_css_modules: bool,
    pub is_virtual: bool,
    pub is_entry: bool,
    pub pathname: PathBuf,
    pub search: String,
    pub params: Vec<(String, String)>,
    pub fragment: Option<String>,
}

impl Default for File {
    fn default() -> Self {
        File {
            path: PathBuf::new(),
            relative_path: PathBuf::new(),
            extname: "".to_string(),
            content: None,
            is_under_node_modules: false,
            is_css_modules: false,
            is_virtual: false,
            is_entry: false,
            pathname: PathBuf::new(),
            search: "".to_string(),
            params: vec![],
            fragment: None,
        }
    }
}

const VIRTUAL: &str = "virtual:";

fn css_source_map_regex() -> &'static Regex {
    static CSS_SOURCE_MAP_REGEXP: OnceLock<Regex> = OnceLock::new();

    CSS_SOURCE_MAP_REGEXP.get_or_init(|| {
        Regex::new(r"/\*# sourceMappingURL=data:application/json;base64,(.*?) \*/").unwrap()
    })
}

impl File {
    pub fn new(path: String, context: Arc<Context>) -> Self {
        let path = PathBuf::from(path);
        // if path exists, it has no search and fragment
        // support ./a#b.ts when a#b.ts is a real file
        // e.g. https://unpkg.com/browse/es5-ext@0.10.64/string/
        let (pathname, search, params, fragment) = if path.exists() {
            (
                path.to_string_lossy().to_string(),
                "".to_string(),
                vec![],
                None,
            )
        } else {
            parse_path(&path.to_string_lossy()).unwrap()
        };
        let pathname = PathBuf::from(pathname);
        let is_virtual = path.starts_with(VIRTUAL) ||
            // TODO: remove this specific logic
            params.iter().any(|(k, _)| k == "asmodule");
        let is_under_node_modules = path.to_string_lossy().contains("node_modules");
        let extname = pathname
            .clone()
            .extension()
            .map(|ext| ext.to_string_lossy().to_string())
            .unwrap_or_default();
        if is_virtual {
            File {
                path: path.clone(),
                relative_path: path,
                is_virtual,
                pathname,
                search,
                params,
                fragment,
                is_under_node_modules,
                extname,
                ..Default::default()
            }
        } else {
            let relative_path = diff_paths(&path, &context.root).unwrap_or(path.clone());
            File {
                is_virtual,
                path,
                relative_path,
                extname,
                is_under_node_modules,
                pathname,
                search,
                params,
                ..Default::default()
            }
        }
    }

    #[allow(dead_code)]
    pub fn with_content(path: String, content: Content, context: Arc<Context>) -> Self {
        let mut file = File::new(path, context);
        file.content = Some(content);
        file
    }

    pub fn new_entry(path: String, context: Arc<Context>) -> Self {
        let mut file = File::new(path, context);
        file.is_entry = true;
        file
    }

    pub fn set_content(&mut self, content: Content) {
        self.content = Some(content);
    }

    pub fn get_content_raw(&self) -> String {
        match &self.content {
            Some(Content::Js(JsContent { content, .. })) | Some(Content::Css(content)) => {
                content.clone()
            }
            Some(Content::Assets(asset)) => asset.content.clone(),
            None => "".to_string(),
        }
    }

    pub fn get_raw_hash(&self) -> u64 {
        let mut hasher: XxHash64 = Default::default();
        if let Some(content) = &self.content {
            match content {
                Content::Js(JsContent { content, .. })
                | Content::Css(content)
                | Content::Assets(Asset { content, .. }) => {
                    // hasher.write_u64(init);
                    hasher.write(content.as_bytes());
                    hasher.finish()
                }
            }
        } else {
            0
        }
    }

    pub fn get_file_stem(&self) -> String {
        self.relative_path
            .file_stem()
            .map(|s| s.to_string_lossy().to_string())
            .unwrap_or_default()
    }

    pub fn get_file_size(&self) -> Result<u64> {
        let metadata = std::fs::metadata(&self.pathname)?;
        Ok(metadata.len())
    }

    pub fn get_base64(&self) -> Result<String> {
        let content = std::fs::read(&self.pathname)?;
        let engine = engine::GeneralPurpose::new(&STANDARD, engine::general_purpose::PAD);
        let content = engine.encode(content);
        let guess = mime_guess::from_path(&self.pathname);
        if let Some(mime) = guess.first() {
            Ok(format!(
                "data:{};base64,{}",
                mime,
                content.replace("\r\n", "")
            ))
        } else {
            Err(anyhow!(FileError::ToBase64Error {
                path: self.path.to_string_lossy().to_string(),
            }))
        }
    }

    pub fn get_content_hash(&self) -> Result<String> {
        let file = std::fs::File::open(&self.pathname)?;
        let len = file.metadata()?.len();
        // Decide on a reasonable buffer size (1MB in this case, fastest will depend on hardware)
        let buf_len = len.min(1_000_000) as usize;
        let mut buf = BufReader::with_capacity(buf_len, file);
        // webpack use md4
        let mut context = md5::Context::new();
        loop {
            // Get a chunk of the file
            let part = buf.fill_buf()?;
            if part.is_empty() {
                break;
            }
            context.consume(part);
            // Tell the buffer that the chunk is consumed
            let part_len = part.len();
            buf.consume(part_len);
        }
        let digest = context.compute();
        let hash = format!("{:x}", digest);
        Ok(hash[0..8].to_string())
    }

    pub fn is_content_jsx(&self) -> bool {
        match &self.content {
            Some(Content::Js(JsContent { is_jsx, .. })) => *is_jsx,
            _ => false,
        }
    }

    pub fn has_param(&self, key: &str) -> bool {
        self.params.iter().any(|(k, _)| k == key)
    }

    pub fn get_source_map_chain(&self, context: Arc<Context>) -> Vec<Vec<u8>> {
        if context.config.devtool.is_none() {
            return vec![];
        }
        let mut chain = vec![];
        match &self.content {
            Some(Content::Css(content)) => {
                if let Some(captures) = css_source_map_regex().captures(content) {
                    let source_map_base64 = captures.get(1).unwrap().as_str().to_string();
                    chain.push(base64_decode(source_map_base64.as_bytes()));
                }
            }
            // TODO: support js source map chain
            Some(Content::Js(_)) => {}
            _ => {}
        }
        chain
    }
}

type PathName = String;
type Search = String;
type Params = Vec<(String, String)>;
type Fragment = Option<String>;
fn parse_path(path: &str) -> Result<(PathName, Search, Params, Fragment)> {
    let base = "http://a.com/";
    let base_url = Url::parse(base)?;
    let full_url = base_url.join(path)?;
    let path = full_url.path().to_string();
    let fragment = full_url.fragment().map(|s| s.to_string());
    let search = full_url.query().unwrap_or("").to_string();
    let query_vec = full_url
        .query_pairs()
        .map(|(k, v)| (k.to_string(), v.to_string()))
        .collect();
    // dir or filename may contains space or other special characters
    // so we need to decode it, e.g. "a%20b" -> "a b"
    let path = percent_decode_str(&path).decode_utf8()?;
    Ok((path.to_string(), search, query_vec, fragment))
}
