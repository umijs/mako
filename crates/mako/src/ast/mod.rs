pub(crate) mod css_ast;
pub(crate) mod error;
pub mod file;
pub(crate) mod js_ast;
#[cfg(test)]
pub mod tests;
pub(crate) mod utils;
