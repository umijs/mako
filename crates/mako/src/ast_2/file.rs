use std::hash::Hasher;
use std::io::{BufRead, BufReader};
use std::path::PathBuf;
use std::sync::Arc;

use mako_core::anyhow::{anyhow, Result};
use mako_core::base64::alphabet::STANDARD;
use mako_core::base64::{engine, Engine};
use mako_core::lazy_static::lazy_static;
use mako_core::pathdiff::diff_paths;
use mako_core::regex::Regex;
use mako_core::thiserror::Error;
use mako_core::twox_hash::XxHash64;
use mako_core::{md5, mime_guess};

use crate::compiler::Context;
use crate::util::base64_decode;

#[derive(Debug, Clone)]
pub struct Asset {
    pub path: String,
    pub content: String,
}

#[derive(Debug, Clone)]
pub enum Content {
    Js(String),
    Css(String),
    // TODO: unify the assets handler
    // it's used in minifish plugin(bundless mode) only
    // and bundle mode will emit assets to context.assets_info
    Assets(Asset),
}

#[derive(Debug, Error)]
enum FileError {
    #[error("To base64 error: {path:?}")]
    ToBase64Error { path: String },
}

#[derive(Debug, Clone)]
pub struct File {
    pub path: PathBuf,
    pub relative_path: PathBuf,
    pub extname: String,
    pub content: Option<Content>,
    pub is_under_node_modules: bool,
    pub is_css_modules: bool,
    pub is_virtual: bool,
    pub is_entry: bool,
    pub pathname: PathBuf,
    pub search: String,
    pub params: Vec<(String, String)>,
}

impl Default for File {
    fn default() -> Self {
        File {
            path: PathBuf::new(),
            relative_path: PathBuf::new(),
            extname: "".to_string(),
            content: None,
            is_under_node_modules: false,
            is_css_modules: false,
            is_virtual: false,
            is_entry: false,
            pathname: PathBuf::new(),
            search: "".to_string(),
            params: vec![],
        }
    }
}

// e.g.
lazy_static! {
    static ref VIRTUAL: String = "virtual:".to_string();
}

lazy_static! {
    static ref CSS_SOURCE_MAP_REGEXP: Regex =
        Regex::new(r"/\*# sourceMappingURL=data:application/json;base64,(.*?) \*/").unwrap();
}

impl File {
    pub fn new(path: String, context: Arc<Context>) -> Self {
        let path = PathBuf::from(path);
        let (pathname, search, params) = parse_path(&path.to_string_lossy()).unwrap();
        let pathname = PathBuf::from(pathname);
        let is_virtual = path.starts_with(&*VIRTUAL) ||
            // TODO: remove this specific logic
            params.iter().any(|(k, _)| k == "asmodule");
        let is_under_node_modules = path.to_string_lossy().contains("node_modules");
        let extname = pathname
            .clone()
            .extension()
            .map(|ext| ext.to_string_lossy().to_string())
            .unwrap_or_default();
        if is_virtual {
            File {
                path: path.clone(),
                relative_path: path,
                is_virtual,
                pathname,
                search,
                params,
                is_under_node_modules,
                extname,
                ..Default::default()
            }
        } else {
            let relative_path = diff_paths(&path, &context.root).unwrap_or(path.clone());
            File {
                is_virtual,
                path,
                relative_path,
                extname,
                is_under_node_modules,
                pathname,
                search,
                params,
                ..Default::default()
            }
        }
    }

    #[allow(dead_code)]
    pub fn with_content(path: String, content: Content, context: Arc<Context>) -> Self {
        let mut file = File::new(path, context);
        file.content = Some(content);
        file
    }

    pub fn new_entry(path: String, context: Arc<Context>) -> Self {
        let mut file = File::new(path, context);
        file.is_entry = true;
        file
    }

    pub fn set_content(&mut self, content: Content) {
        self.content = Some(content);
    }

    pub fn get_content_raw(&self) -> String {
        match &self.content {
            Some(Content::Js(content)) | Some(Content::Css(content)) => content.clone(),
            Some(Content::Assets(asset)) => asset.content.clone(),
            None => "".to_string(),
        }
    }

    pub fn get_raw_hash(&self, init: u64) -> u64 {
        let mut hasher: XxHash64 = Default::default();
        if let Some(content) = &self.content {
            match content {
                Content::Js(content)
                | Content::Css(content)
                | Content::Assets(Asset { content, .. }) => {
                    hasher.write_u64(init);
                    hasher.write(content.as_bytes());
                    hasher.finish()
                }
            }
        } else {
            0
        }
    }

    pub fn get_file_stem(&self) -> String {
        self.relative_path
            .file_stem()
            .map(|s| s.to_string_lossy().to_string())
            .unwrap_or_default()
    }

    pub fn get_file_size(&self) -> Result<u64> {
        let metadata = std::fs::metadata(&self.path)?;
        Ok(metadata.len())
    }

    pub fn get_base64(&self) -> Result<String> {
        let content = std::fs::read(&self.path)?;
        let engine = engine::GeneralPurpose::new(&STANDARD, engine::general_purpose::PAD);
        let content = engine.encode(content);
        let guess = mime_guess::from_path(&self.path);
        if let Some(mime) = guess.first() {
            Ok(format!(
                "data:{};base64,{}",
                mime,
                content.replace("\r\n", "")
            ))
        } else {
            Err(anyhow!(FileError::ToBase64Error {
                path: self.path.to_string_lossy().to_string(),
            }))
        }
    }

    pub fn get_content_hash(&self) -> Result<String> {
        let file = std::fs::File::open(&self.path)?;
        let len = file.metadata()?.len();
        // Decide on a reasonable buffer size (1MB in this case, fastest will depend on hardware)
        let buf_len = len.min(1_000_000) as usize;
        let mut buf = BufReader::with_capacity(buf_len, file);
        // webpack use md4
        let mut context = md5::Context::new();
        loop {
            // Get a chunk of the file
            let part = buf.fill_buf()?;
            if part.is_empty() {
                break;
            }
            context.consume(part);
            // Tell the buffer that the chunk is consumed
            let part_len = part.len();
            buf.consume(part_len);
        }
        let digest = context.compute();
        let hash = format!("{:x}", digest);
        Ok(hash[0..8].to_string())
    }

    pub fn has_param(&self, key: &str) -> bool {
        self.params.iter().any(|(k, _)| k == key)
    }

    pub fn get_source_map_chain(&self, context: Arc<Context>) -> Vec<Vec<u8>> {
        if context.config.devtool.is_none() {
            return vec![];
        }
        let mut chain = vec![];
        match &self.content {
            Some(Content::Css(content)) => {
                if let Some(captures) = CSS_SOURCE_MAP_REGEXP.captures(content) {
                    let source_map_base64 = captures.get(1).unwrap().as_str().to_string();
                    chain.push(base64_decode(source_map_base64.as_bytes()));
                }
            }
            // TODO: support js source map chain
            Some(Content::Js(_)) => {}
            _ => {}
        }
        chain
    }
}

type PathName = String;
type Search = String;
type Params = Vec<(String, String)>;
fn parse_path(path: &str) -> Result<(PathName, Search, Params)> {
    let mut iter = path.split('?');
    let path = iter.next().unwrap();
    let query = iter.next().unwrap_or("");
    let mut query_vec = vec![];
    for pair in query.split('&') {
        if pair.contains('=') {
            let mut it = pair.split('=').take(2);
            let kv = match (it.next(), it.next()) {
                (Some(k), Some(v)) => (k.to_string(), v.to_string()),
                _ => continue,
            };
            query_vec.push(kv);
        } else if !pair.is_empty() {
            query_vec.push((pair.to_string(), "".to_string()));
        }
    }
    let search = if query.is_empty() {
        "".to_string()
    } else {
        format!("?{}", query)
    };
    Ok((path.to_string(), search, query_vec))
}
