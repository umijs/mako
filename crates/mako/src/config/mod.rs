#[allow(clippy::module_inception)]
mod config;
pub use config::*;
