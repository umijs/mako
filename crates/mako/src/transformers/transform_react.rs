use std::sync::Arc;

use mako_core::swc_common::sync::Lrc;
use mako_core::swc_common::{chain, Mark, SourceMap};
use mako_core::swc_ecma_ast::Module;
use mako_core::swc_ecma_transforms_react::{react, Options, RefreshOptions, Runtime};
use mako_core::swc_ecma_visit::{VisitMut, VisitMutWith};

use crate::ast::build_js_ast;
use crate::ast_2::file::File;
use crate::compiler::Context;
use crate::config::{Mode, ReactRuntimeConfig};

pub struct PrefixCode {
    pub code: String,
    pub context: Arc<Context>,
}

pub fn mako_react(
    cm: Lrc<SourceMap>,
    context: Arc<Context>,
    file: &File,
    top_level_mark: &Mark,
    unresolved_mark: &Mark,
) -> Box<dyn VisitMut> {
    let is_dev = matches!(context.config.mode, Mode::Development);
    let is_browser = matches!(context.config.platform, crate::config::Platform::Browser);
    let use_refresh = is_dev
        && context.args.watch
        && context.config.hmr.is_some()
        && !file.is_under_node_modules
        && is_browser;

    let is_jsx = file.extname == "jsx"
        || file.extname == "js"
        || file.extname == "ts"
        || file.extname == "tsx"
        || file.extname == "svg";

    if !is_jsx {
        return if file.is_entry && use_refresh {
            Box::new(chain!(react_refresh_inject_runtime_only(&context), noop()))
        } else {
            Box::new(noop())
        };
    }

    let runtime = if matches!(context.config.react.runtime, ReactRuntimeConfig::Automatic) {
        Runtime::Automatic
    } else {
        Runtime::Classic
    };

    let origin_comments = context.meta.script.origin_comments.read().unwrap();
    let visit = react(
        cm,
        Some(origin_comments.get_swc_comments().clone()),
        Options {
            import_source: Some(context.config.react.import_source.clone()),
            pragma: Some(context.config.react.pragma.clone()),
            pragma_frag: Some(context.config.react.pragma_frag.clone()),
            // support react 17 + only
            runtime: Some(runtime),
            development: Some(is_dev),
            // to avoid throw error for svg namespace element
            throw_if_namespace: Some(false),
            refresh: if use_refresh {
                Some(RefreshOptions::default())
            } else {
                None
            },
            ..Default::default()
        },
        *top_level_mark,
        *unresolved_mark,
    );
    if use_refresh {
        Box::new(chain!(
            visit,
            react_refresh_module_prefix(&context),
            react_refresh_module_postfix(&context)
        ))
    } else {
        Box::new(visit)
    }
}

impl VisitMut for PrefixCode {
    fn visit_mut_module(&mut self, module: &mut Module) {
        let post_code_snippet_module =
            build_js_ast("_pre_code.js", &self.code, &self.context).unwrap();
        module.body.splice(0..0, post_code_snippet_module.ast.body);

        module.visit_mut_children_with(self);
    }
}

pub struct PostfixCode {
    code: String,
    context: Arc<Context>,
}

impl VisitMut for PostfixCode {
    fn visit_mut_module(&mut self, module: &mut Module) {
        let post_code_snippet_module =
            build_js_ast("_post_code.js", &self.code, &self.context).unwrap();
        module.body.extend(post_code_snippet_module.ast.body);

        module.visit_mut_children_with(self);
    }
}

pub fn react_refresh_module_prefix(context: &std::sync::Arc<Context>) -> Box<dyn VisitMut> {
    Box::new(PrefixCode {
        context: context.clone(),
        code: r#"
import * as RefreshRuntime from 'react-refresh';
var prevRefreshReg;
var prevRefreshSig;

prevRefreshReg = self.$RefreshReg$;
prevRefreshSig = self.$RefreshSig$;
self.$RefreshReg$ = (type, id) => {
  RefreshRuntime.register(type, module.id + id);
};
self.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;
"#
        .to_string(),
    })
}

pub fn react_refresh_inject_runtime_only(context: &std::sync::Arc<Context>) -> Box<dyn VisitMut> {
    Box::new(PrefixCode {
        context: context.clone(),
        code: r#"
import 'react-refresh';
"#
        .to_string(),
    })
}

pub fn react_refresh_module_postfix(context: &Arc<Context>) -> Box<dyn VisitMut> {
    Box::new(PostfixCode {
        context: context.clone(),
        code: r#"
self.$RefreshReg$ = prevRefreshReg;
self.$RefreshSig$ = prevRefreshSig;
function $RefreshIsReactComponentLike$(moduleExports) {
  if (RefreshRuntime.isLikelyComponentType(moduleExports.default || moduleExports)) {
    return true;
  }
  for (var key in moduleExports) {
    if (RefreshRuntime.isLikelyComponentType(moduleExports[key])) {
      return true;
    }
  }
  return false;
}
if ($RefreshIsReactComponentLike$(module.exports)) {
    module.meta.hot.accept();
    RefreshRuntime.performReactRefresh();
}
"#
        .to_string(),
    })
}

struct NoopVisitor;

impl VisitMut for NoopVisitor {
    fn visit_mut_module(&mut self, _: &mut Module) {
        // Do nothing.
    }
}

fn noop() -> Box<dyn VisitMut> {
    Box::new(NoopVisitor)
}

// #[cfg(test)]
// mod tests {

//     use std::sync::Arc;

//     use mako_core::swc_common::{chain, Mark, GLOBALS};
//     use mako_core::swc_ecma_transforms::resolver;
//     use mako_core::swc_ecma_visit::VisitMut;

//     use crate::assert_display_snapshot;
//     use crate::ast::build_js_ast;
//     use crate::ast_2::file::File;
//     use crate::compiler::{Args, Context};
//     use crate::test_helper::transform_ast_with;
//     use crate::transformers::transform_react::mako_react;

//     struct TransformTask {
//         code: String,
//         path: String,
//         is_entry: bool,
//     }

//     #[test]
//     pub fn entry_with_react_refresh() {
//         assert_display_snapshot!(transform(TransformTask {
//             is_entry: true,
//             path: "index.js".to_string(),
//             code: "console.log('entry');".to_string()
//         }));
//     }

//     #[test]
//     pub fn node_modules_with_react_refresh() {
//         assert_display_snapshot!(transform(TransformTask {
//             code: "console.log('in node modules');".to_string(),
//             is_entry: false,
//             path: "project/node_modules/pkg/index.js".to_string()
//         }));
//     }

//     #[test]
//     pub fn normal_module_with_react_refresh() {
//         assert_display_snapshot!(transform(TransformTask {
//             code: "export default function R(){return <h1></h1>}".to_string(),
//             is_entry: false,
//             path: "index.jsx".to_string()
//         }));
//     }

//     #[test]
//     pub fn svgr_with_namespace() {
//         assert_display_snapshot!(transform(TransformTask {
//             // part of jsoneditor/dist/img/jsoneditor-icons.svg
//             code: r#"const SvgComponent = (props) => (
//     <svg
//         xmlns:dc="http://purl.org/dc/elements/1.1/"
//         xmlns:cc="http://creativecommons.org/ns#"
//         xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
//         xmlns:svg="http://www.w3.org/2000/svg"
//         xmlns="http://www.w3.org/2000/svg"
//         xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
//         xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
//         width={240}
//         height={144}
//         id="svg4136"
//         inkscape:version="0.91 r13725"
//         sodipodi:docname="jsoneditor-icons.svg"
//         {...props}
//     >
//         <metadata id="metadata4148">
//             <rdf:RDF></rdf:RDF>
//         </metadata>
//     </svg>
// )"#
//             .to_string(),
//             is_entry: false,
//             path: "index.jsx".to_string()
//         }));
//     }

//     fn transform(task: TransformTask) -> String {
//         let context: Arc<Context> = Arc::new(Context {
//             args: Args { watch: true },
//             ..Default::default()
//         });

//         GLOBALS.set(&context.meta.script.globals, || {
//             let mut ast = build_js_ast("index.jsx", &task.code, &context).unwrap();

//             let file = if task.is_entry {
//                 File::new_entry(task.path.clone(), context)
//             } else {
//                 File::new(task.path.clone(), context)
//             };
//             let mut visitor: Box<dyn VisitMut> = Box::new(chain!(
//                 resolver(Mark::new(), Mark::new(), false),
//                 mako_react(
//                     Default::default(),
//                     &context,
//                     &file,
//                     &Mark::new(),
//                     &Mark::new(),
//                 )
//             ));

//             transform_ast_with(&mut ast.ast, &mut visitor, &context.meta.script.cm)
//         })
//     }
// }
