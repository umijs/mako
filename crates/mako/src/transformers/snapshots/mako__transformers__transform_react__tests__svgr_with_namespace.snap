---
source: crates/mako/src/transformers/transform_react.rs
expression: "transform(TransformTask {\n        code: r#\"const SvgComponent = (props) => (\n    <svg\n        xmlns:dc=\"http://purl.org/dc/elements/1.1/\"\n        xmlns:cc=\"http://creativecommons.org/ns#\"\n        xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\"\n        xmlns:svg=\"http://www.w3.org/2000/svg\"\n        xmlns=\"http://www.w3.org/2000/svg\"\n        xmlns:sodipodi=\"http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd\"\n        xmlns:inkscape=\"http://www.inkscape.org/namespaces/inkscape\"\n        width={240}\n        height={144}\n        id=\"svg4136\"\n        inkscape:version=\"0.91 r13725\"\n        sodipodi:docname=\"jsoneditor-icons.svg\"\n        {...props}\n    >\n        <metadata id=\"metadata4148\">\n            <rdf:RDF></rdf:RDF>\n        </metadata>\n    </svg>\n)\"#.to_string(),\n        is_entry: false,\n        path: \"index.jsx\".to_string(),\n    })"
---
import * as RefreshRuntime from 'react-refresh';
var prevRefreshReg;
var prevRefreshSig;
prevRefreshReg = self.$RefreshReg$;
prevRefreshSig = self.$RefreshSig$;
self.$RefreshReg$ = (type, id)=>{
    RefreshRuntime.register(type, module.id + id);
};
self.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;
import { jsxDEV as _jsxDEV } from "react/jsx-dev-runtime";
const SvgComponent = (props)=>_jsxDEV("svg", {
        "xmlns:dc": "http://purl.org/dc/elements/1.1/",
        "xmlns:cc": "http://creativecommons.org/ns#",
        "xmlns:rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
        "xmlns:svg": "http://www.w3.org/2000/svg",
        xmlns: "http://www.w3.org/2000/svg",
        "xmlns:sodipodi": "http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd",
        "xmlns:inkscape": "http://www.inkscape.org/namespaces/inkscape",
        width: 240,
        height: 144,
        id: "svg4136",
        "inkscape:version": "0.91 r13725",
        "sodipodi:docname": "jsoneditor-icons.svg",
        ...props,
        children: _jsxDEV("metadata", {
            id: "metadata4148",
            children: _jsxDEV("rdf:RDF", {}, void 0, false, {
                fileName: "<<jsx-config-pragmaFrag.js>>",
                lineNumber: 1,
                columnNumber: 648
            }, this)
        }, void 0, false, {
            fileName: "<<jsx-config-pragmaFrag.js>>",
            lineNumber: 1,
            columnNumber: 607
        }, this)
    }, void 0, false, {
        fileName: "<<jsx-config-pragmaFrag.js>>",
        lineNumber: 1,
        columnNumber: 19
    }, this);
_c = SvgComponent;
var _c;
$RefreshReg$(_c, "SvgComponent");
self.$RefreshReg$ = prevRefreshReg;
self.$RefreshSig$ = prevRefreshSig;
function $RefreshIsReactComponentLike$(moduleExports) {
    if (RefreshRuntime.isLikelyComponentType(moduleExports.default || moduleExports)) {
        return true;
    }
    for(var key in moduleExports){
        if (RefreshRuntime.isLikelyComponentType(moduleExports[key])) {
            return true;
        }
    }
    return false;
}
if ($RefreshIsReactComponentLike$(module.exports)) {
    module.meta.hot.accept();
    RefreshRuntime.performReactRefresh();
}
